src/block_cache/block_chunkcache.c

```diff
@@ -0,0 +1,552 @@
+                left_to_read -= size_copied;
+
+                WT_STAT_CONN_INCR(session, chunk_cache_hits);
+                break;
+            }
+        }
+        /* The chunk is not cached. Allocate space for it. Prepare for reading it from storage. */
+        if (!chunk_cached) {
+            if ((ret = __chunkcache_alloc_chunk(
+                   session, offset + (wt_off_t)already_read, block, &hash_id, &chunk)) != 0) {
+                __wt_spin_unlock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
+                return (ret);
+            }
+            /*
+             * Insert the invalid chunk into the bucket before releasing the lock and doing I/O.
+             * This way we avoid two threads trying to cache the same chunk.
+             */
+            TAILQ_INSERT_HEAD(WT_BUCKET_CHUNKS(chunkcache, bucket_id), chunk, next_chunk);
+            __wt_spin_unlock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
+
+            /* Read the new chunk. Only one thread would be caching the new chunk. */
+            if ((ret = __wt_read(session, block->fh, chunk->chunk_offset, chunk->chunk_size,
+                   chunk->chunk_memory)) != 0) {
+                __wt_spin_lock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
+                TAILQ_REMOVE(WT_BUCKET_CHUNKS(chunkcache, bucket_id), chunk, next_chunk);
+                __wt_spin_unlock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
+                __chunkcache_free_chunk(session, chunk);
+                WT_STAT_CONN_INCR(session, chunk_cache_io_failed);
+                return (ret);
+            }
+
+            /*
+             * Mark chunk as valid. The only thread that could be executing this code is the thread
+             * that won the race and inserted this (invalid) chunk into the hash table. This thread
+             * has now read the chunk, while any other threads that were looking for the same chunk
+             * would be spin-waiting for this chunk to become valid. The current thread will mark
+             * the chunk as valid, and any waiters will unblock and proceed reading it.
+             */
+            (void)__wt_atomic_addv32(&chunk->valid, 1);
+
+            /* Insert the new chunk into the LRU list */
+            __wt_spin_lock(session, &chunkcache->chunkcache_lru_lock);
+            TAILQ_INSERT_HEAD(&chunkcache->chunkcache_lru_list, chunk, next_lru_item);
+            __wt_spin_unlock(session, &chunkcache->chunkcache_lru_lock);
+
+            __wt_verbose(session, WT_VERB_CHUNKCACHE,
+              "insert: %s(%u), offset=%" PRId64 ", size=%lu", (char *)block->name, objectid,
+              chunk->chunk_offset, chunk->chunk_size);
+            goto retry;
+        }
+    }
+    return (0);
+}
+
+/*
+ * __wt_chunkcache_remove --
+ *     Remove the chunk containing an outdated block.
+ */
+void
+__wt_chunkcache_remove(
+  WT_SESSION_IMPL *session, WT_BLOCK *block, uint32_t objectid, wt_off_t offset, uint32_t size)
```

Comment:
@ddanderson , imagine the following scenario. The block manager frees the block, makes a call to remove, but we don't invalidate the chunk. Then the block manager reuses that block for something else, writes it, but we don't know about it. Then upon a read, we may return an invalid block. To prevent this from happening, we need to invalidate the entire chunk (or make the code more complicated to invalidate parts of the chunk). In the future we will enforce the checks that the chunk cache only caches read-only objects, and this will be less relevant -- the "remove" call will never be used and we won't have to support it.

Reply:
Got it, thanks!